File: ring-android/app/src/main/java/cx/ring/settings/pluginssettings/PluginsListSettingsFragment.java
package cx.ring.settings.pluginssettings;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.material.snackbar.Snackbar;

import java.io.File;
import java.io.IOException;
import java.util.List;

import cx.ring.R;
import cx.ring.account.JamiAccountSummaryFragment;
import cx.ring.client.HomeActivity;
import cx.ring.daemon.Ringservice;
import cx.ring.databinding.FragPluginsListSettingsBinding;
import cx.ring.plugins.PluginUtils;
import cx.ring.utils.AndroidFileUtils;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;


public class PluginsListSettingsFragment extends Fragment implements PluginsListAdapter.PluginListItemListener {

    public static final String TAG = PluginsListSettingsFragment.class.getSimpleName();

    private final OnBackPressedCallback mOnBackPressedCallback = new OnBackPressedCallback(false) {
        @Override
        public void handleOnBackPressed() {
            mOnBackPressedCallback.setEnabled(false);
            JamiAccountSummaryFragment fragment = (JamiAccountSummaryFragment) getParentFragment();
            if (fragment != null) {
                fragment.popBackStack();
            }
        }
    };

    private static final int ARCHIVE_REQUEST_CODE = 42;

    private FragPluginsListSettingsBinding binding;
    private PluginsListAdapter mAdapter;
    private final CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragPluginsListSettingsBinding.inflate(inflater, container, false);

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        binding.pluginsList.setHasFixedSize(true);

        // specify an adapter (see also next example)
        mAdapter = new PluginsListAdapter(PluginUtils.getInstalledPlugins(binding.pluginsList.getContext()), this);
        binding.pluginsList.setAdapter(mAdapter);

        //Fab
        binding.pluginsListSettingsFab.setOnClickListener(view -> {
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("*/*");
            startActivityForResult(intent, ARCHIVE_REQUEST_CODE);
        });

        return binding.getRoot();
    }

    @Override
    public void onResume() {
        mOnBackPressedCallback.setEnabled(true);
        super.onResume();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        requireActivity().getOnBackPressedDispatcher().addCallback(this, mOnBackPressedCallback);
    }

    /**
     * Implements PluginListItemListener.onPluginItemClicked which is called when we click on
     * a plugin list item
     * @param pluginDetails instance of a plugin details that is sent to PluginSettingsFragment
     */
    @Override
    public void onPluginItemClicked(PluginDetails pluginDetails) {
        ((HomeActivity) requireActivity()).gotToPluginSettings(pluginDetails);
    }

    /**
     * Implements PluginListItemListener.onPluginEnabled which is called when the checkbox
     * associated with the plugin list item is called
     * @param pluginDetails instance of a plugin details that is sent to PluginSettingsFragment
     */
    @Override
    public void onPluginEnabled(PluginDetails pluginDetails) {

        if(pluginDetails.isEnabled()) {
            pluginDetails.setEnabled(PluginUtils.loadPlugin(pluginDetails.getRootPath()));
        } else {
            PluginUtils.unloadPlugin(pluginDetails.getRootPath());
        }
        String status = pluginDetails.isEnabled()?"ON":"OFF";
        Toast.makeText(requireContext(), pluginDetails.getName() + " " + status,
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == ARCHIVE_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            if(data != null) {
                Uri uri = data.getData();
                if (uri != null) {
                    installPluginFromUri(uri, false);
                }
            }
         }
    }

    private String installPluginFile(File pluginFile, boolean force) throws IOException{
        int i = Ringservice.installPlugin(pluginFile.getAbsolutePath(), force);
        if (!pluginFile.delete()) {
            Log.e(TAG,"Plugin Jpl file in the cache not freed");
        }
        switch (i) {
            case 0:
                return pluginFile.getName();
            case 100:
                throw new IOException(getResources()
                        .getString(R.string.plugin_same_version_exception,
                                pluginFile.getName()));
            case 200:
                throw new IOException(getResources()
                        .getString(R.string.plugin_recent_version_exception,
                                pluginFile.getName()));
            default:
                throw new IOException(getResources()
                        .getString(R.string.plugin_install_failure,
                                pluginFile.getName()));
        }
    }

    private void installPluginFromUri(Uri uri, boolean force) {
        mCompositeDisposable.add(AndroidFileUtils.getCacheFile(requireContext(), uri)
                .observeOn(AndroidSchedulers.mainThread())
                .map(file -> installPluginFile(file, force))
                .subscribe(filename -> {
                        String[] plugin = filename.split(".jpl");
                        List<PluginDetails> availablePlugins = PluginUtils.getInstalledPlugins(requireContext());
                        for (PluginDetails availablePlugin : availablePlugins) {
                            if (availablePlugin.getName().equals(plugin[0])) {
                                availablePlugin.setEnabled(true);
                                onPluginEnabled(availablePlugin);
                            }
                        }
                        mAdapter.updatePluginsList(PluginUtils.getInstalledPlugins(requireContext()));
                        Toast.makeText(requireContext(), "Plugin: " + filename + " successfully installed", Toast.LENGTH_LONG).show();
                        mCompositeDisposable.dispose();
                    }, e -> {
                        if (binding != null) {
                            Snackbar sb = Snackbar.make(binding.listLayout, "" + e.getMessage(), Snackbar.LENGTH_LONG);
                            sb.setAction(R.string.plugin_force_install, v -> installPluginFromUri(uri, true));
                            sb.show();
                        }
                    }));
    }

    @Override
    public void onDestroyView() {
        binding = null;
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mCompositeDisposable.dispose();
    }
}

File: ring-android/app/src/main/java/cx/ring/settings/pluginssettings/PluginsListSettingsFragment.kt
package cx.ring.settings.pluginssettings

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar
import cx.ring.R
import cx.ring.account.AccountEditionFragment
import cx.ring.client.HomeActivity
import cx.ring.databinding.FragPluginsListSettingsBinding
import cx.ring.plugins.PluginUtils.getInstalledPlugins
import cx.ring.plugins.PluginUtils.loadPlugin
import cx.ring.plugins.PluginUtils.unloadPlugin
import cx.ring.settings.pluginssettings.PluginsListAdapter.PluginListItemListener
import cx.ring.utils.AndroidFileUtils.getCacheFile
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import net.jami.daemon.JamiService
import java.io.File
import java.io.IOException

class PluginsListSettingsFragment : Fragment(), PluginListItemListener {
    private var binding: FragPluginsListSettingsBinding? = null
    private var mAdapter: PluginsListAdapter? = null
    private val mCompositeDisposable = CompositeDisposable()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragPluginsListSettingsBinding.inflate(inflater, container, false)
        val accountId = requireArguments().getString(AccountEditionFragment.ACCOUNT_ID_KEY)!!

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        binding!!.pluginsList.setHasFixedSize(true)

        mAdapter = PluginsListAdapter(getInstalledPlugins(binding!!.pluginsList.context), this, accountId)
        binding!!.pluginsList.adapter = mAdapter

        //Fab
        if (accountId.isEmpty()) {
            binding!!.pluginsListSettingsFab.visibility = View.VISIBLE
            binding!!.pluginsListSettingsFab.setOnClickListener {
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "*/*"
                startActivityForResult(intent, ARCHIVE_REQUEST_CODE)
            }
        }
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding!!.toolbar.setTitle(R.string.menu_item_plugin_list)
        binding!!.toolbar.setNavigationIcon(androidx.appcompat.R.drawable.abc_ic_ab_back_material)
        binding!!.toolbar.setNavigationOnClickListener {
            activity?.onBackPressed()
        }
    }

    /**
     * Implements PluginListItemListener.onPluginItemClicked which is called when we click on
     * a plugin list item
     * @param pluginDetails instance of a plugin details that is sent to PluginSettingsFragment
     */
    override fun onPluginItemClicked(pluginDetails: PluginDetails) {
        (requireActivity() as HomeActivity).gotToPluginSettings(pluginDetails)
    }

    /**
     * Implements PluginListItemListener.onPluginEnabled which is called when the checkbox
     * associated with the plugin list item is called
     * @param pluginDetails instance of a plugin details that is sent to PluginSettingsFragment
     */
    override fun onPluginEnabled(pluginDetails: PluginDetails) {
        if (pluginDetails.isEnabled) {
            pluginDetails.isEnabled = loadPlugin(pluginDetails.rootPath)
        } else {
            unloadPlugin(pluginDetails.rootPath)
        }
        val status = if (pluginDetails.isEnabled) "ON" else "OFF"
        Toast.makeText(
            requireContext(), pluginDetails.name + " " + status,
            Toast.LENGTH_SHORT
        ).show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == ARCHIVE_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            if (data != null) {
                val uri = data.data
                if (uri != null) {
                    installPluginFromUri(uri, false)
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun installPluginFile(pluginFile: File, force: Boolean): String {
        val i = JamiService.installPlugin(pluginFile.absolutePath, force)
        if (!pluginFile.delete()) {
            Log.e(TAG, "Plugin Jpl file in the cache not freed")
        }
        return when (i) {
            0 -> pluginFile.name
            100 -> throw IOException(
                resources
                    .getString(
                        R.string.plugin_same_version_exception,
                        pluginFile.name
                    )
            )
            200 -> throw IOException(
                resources
                    .getString(
                        R.string.plugin_recent_version_exception,
                        pluginFile.name
                    )
            )
            else -> throw IOException(
                resources
                    .getString(
                        R.string.plugin_install_failure,
                        pluginFile.name
                    )
            )
        }
    }

    private fun installPluginFromUri(uri: Uri, force: Boolean) {
        mCompositeDisposable.add(
            getCacheFile(requireContext(), uri)
                .observeOn(AndroidSchedulers.mainThread())
                .map { file: File -> installPluginFile(file, force) }
                .subscribe({ filename: String ->
                    val plugin = filename.split(".jpl".toRegex()).toTypedArray()
                    val availablePlugins = getInstalledPlugins(requireContext())
                    for (availablePlugin in availablePlugins) {
                        if (availablePlugin.name == plugin[0]) {
                            availablePlugin.isEnabled = true
                            onPluginEnabled(availablePlugin)
                        }
                    }
                    mAdapter!!.updatePluginsList(getInstalledPlugins(requireContext()))
                    Toast.makeText(requireContext(), "Plugin: $filename successfully installed", Toast.LENGTH_LONG)
                        .show()
                    mCompositeDisposable.dispose()
                }) { e: Throwable ->
                    if (binding != null) {
                        val sb = Snackbar.make(binding!!.listLayout, "" + e.message, Snackbar.LENGTH_LONG)
                        sb.setAction(R.string.plugin_force_install) { v: View? -> installPluginFromUri(uri, true) }
                        sb.show()
                    }
                })
    }

    override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
        mCompositeDisposable.dispose()
    }

    companion object {
        val TAG = PluginsListSettingsFragment::class.java.simpleName
        private const val ARCHIVE_REQUEST_CODE = 42

        fun newInstance(accountId: String?): PluginsListSettingsFragment {
            val fragment = PluginsListSettingsFragment()
            fragment.arguments = Bundle().apply { putString(AccountEditionFragment.ACCOUNT_ID_KEY, accountId) }
            return fragment
        }
    }
}